core/src/main/java/org/geysermc/geyser/registry/loader/ResourcePackLoader.java

```diff
@@ -0,0 +1,157 @@
+ */
+public class ResourcePackLoader implements RegistryLoader<Path, Map<String, ResourcePack>> {
+
+    static final PathMatcher PACK_MATCHER = FileSystems.getDefault().getPathMatcher("glob:**.{zip,mcpack}");
+
+    /**
+     * Loop through the packs directory and locate valid resource pack files
+     */
+    @Override
+    public Map<String, ResourcePack> load(Path directory) {
+        Map<String, ResourcePack> packMap = new HashMap<>();
+
+        if (!Files.exists(directory)) {
+            try {
+                Files.createDirectory(directory);
+            } catch (IOException e) {
+                GeyserImpl.getInstance().getLogger().error("Could not create packs directory", e);
+            }
+
+            return packMap; // As we just created the directory it will be empty
+        }
+
+        List<Path> resourcePacks;
+        try {
+            resourcePacks = Files.walk(directory)
+                    .filter(PACK_MATCHER::matches)
+                    .collect(Collectors.toCollection(ArrayList::new)); // toList() does not guarantee mutability
+        } catch (Exception e) {
+            GeyserImpl.getInstance().getLogger().error("Could not list packs directory", e);
+            return packMap;
+        }
+
+        GeyserLoadResourcePacksEvent event = new GeyserLoadResourcePacksEvent(resourcePacks);
+        GeyserImpl.getInstance().eventBus().fire(event);
+
+        for (Path path : event.resourcePacks()) {
+            try {
+                GeyserResourcePack pack = readPack(path);
+                packMap.put(pack.manifest().header().uuid().toString(), pack);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+        return packMap;
+    }
+
+    /**
+     * Reads a resource pack at the given file. Also searches for a file in the same directory, with the same name
+     * but suffixed by ".key", containing the content key. If such file does not exist, no content key is stored.
+     *
+     * @param path the file to read from, in ZIP format
+     * @return a {@link ResourcePack} representation
+     * @throws IllegalArgumentException if the pack manifest was invalid or there was any processing exception
+     */
+    public static GeyserResourcePack readPack(Path path) throws IllegalArgumentException {
+        GeyserResourcePack pack;
+
+        try (ZipFile zip = new ZipFile(path.toFile());
+             Stream<? extends ZipEntry> stream = zip.stream()) {
+
+            ResourcePackManifest manifest = readManifest(zip, "manifest.json");
+            if (manifest == null || manifest.header().uuid() == null) {
+
+                // Outdated naming?
+                manifest = readManifest(zip, "pack_manifest.json");
+                if (manifest == null || manifest.header().uuid() == null) {
+                    throw new IllegalArgumentException(path.getFileName() + " does not contain a valid pack_manifest.json or manifest.json");
```

Comment:
Kind of ugly because this will be caught and rethrown below... not sure if it matters that much.